File: python/src/moltchess/llm/board.py
from __future__ import annotations

import chess


def legal_moves_from_fen(fen: str) -> tuple[list[str], list[str], dict[str, str]]:
    board = chess.Board(fen)
    sans: list[str] = []
    ucis: list[str] = []
    uci_to_san: dict[str, str] = {}
    for move in board.legal_moves:
        uci = move.uci()
        san = board.san(move)
        sans.append(san)
        ucis.append(uci)
        uci_to_san[uci] = san
    return sans, ucis, uci_to_san


def first_legal_fallback(sans: list[str], ucis: list[str], uci_to_san: dict[str, str]) -> tuple[str, str]:
    if ucis:
        u = ucis[0]
        return uci_to_san.get(u, sans[0] if sans else ""), u
    raise ValueError("No legal moves in position")

File: javascript/src/llm/board.ts
import { Chess } from "chess.js";

export interface LegalMovesBundle {
  legalSans: string[];
  legalUci: string[];
  uciToSan: Map<string, string>;
}

export function legalMovesFromFen(fen: string): LegalMovesBundle {
  const chess = new Chess(fen);
  const verbose = chess.moves({ verbose: true });
  const legalSans: string[] = [];
  const legalUci: string[] = [];
  const uciToSan = new Map<string, string>();
  for (const m of verbose) {
    const uci = m.from + m.to + (m.promotion ?? "");
    legalSans.push(m.san);
    legalUci.push(uci);
    uciToSan.set(uci, m.san);
  }
  return { legalSans, legalUci, uciToSan };
}

export function firstLegalFallback(bundle: LegalMovesBundle): { san: string; uci: string } {
  const uci = bundle.legalUci[0];
  if (!uci) {
    throw new Error("No legal moves in position");
  }
  const san = bundle.uciToSan.get(uci) ?? bundle.legalSans[0] ?? "";
  return { san, uci };
}
